xcommon.cmake

```diff
@@ -794,48 +826,63 @@ function(XMOS_REGISTER_DEPS)
 
 endfunction()
 
+# Takes as input the LIB_NAME or archive name from LIB_ARCHIVES, and removes the toolchain's
+# static library prefix if necessary to avoid the resulting archive being called liblib_foo
+function(get_archive_name IN_NAME RET_NAME)
+    if(CMAKE_STATIC_LIBRARY_PREFIX)
+        string(REGEX REPLACE "^${CMAKE_STATIC_LIBRARY_PREFIX}" "" archive_name ${IN_NAME})
+    else()
+        set(archive_name ${IN_NAME})
+    endif()
+    set(${RET_NAME} ${archive_name} PARENT_SCOPE)
+endfunction()
+
 ## Registers a static library target
 function(XMOS_STATIC_LIBRARY)
     message(STATUS "Configuring static library: ${LIB_NAME}")
 
-    if(NOT BUILD_NATIVE)
-        list(LENGTH LIB_ARCH num_arch)
-        if(${num_arch} LESS 1)
-            # If architecture not specified, assume xs3a
-            set(LIB_ARCH "xs3a")
-        endif()
-    else()
-        set(LIB_ARCH "${CMAKE_HOST_SYSTEM_PROCESSOR}")
-    endif()
-    message(VERBOSE "Building for architecture: ${LIB_ARCH}")
-
-    glob_srcs("LIB" ${CMAKE_CURRENT_SOURCE_DIR} libsrc)
+    glob_srcs("LIB_ARCHIVE" ${CMAKE_CURRENT_SOURCE_DIR} libsrc)
 
-    set(archive_name ${LIB_NAME})
-    if(NOT ${LIB_NAME} MATCHES "^lib")
-        set(archive_name "lib${LIB_NAME}")
+    if(NOT LIB_ARCHIVES)
+        set(LIB_ARCHIVES ${LIB_NAME})
+        if(NOT LIB_ARCHIVE_COMPILER_FLAGS_${LIB_NAME})
+            set(LIB_ARCHIVE_COMPILER_FLAGS_${LIB_NAME} "${LIB_ARCHIVE_COMPILER_FLAGS}")
+        endif()
     endif()
 
     set(APP_BUILD_TARGETS "")
-    foreach(lib_arch ${LIB_ARCH})
-        add_library(${LIB_NAME}-${lib_arch} STATIC)
-        set_property(TARGET ${LIB_NAME}-${lib_arch} PROPERTY VERSION ${LIB_VERSION})
-        target_sources(${LIB_NAME}-${lib_arch} PRIVATE ${LIB_XC_SRCS} ${LIB_CXX_SRCS} ${LIB_ASM_SRCS} ${LIB_C_SRCS})
-        target_include_directories(${LIB_NAME}-${lib_arch} PRIVATE ${LIB_INCLUDES})
+    foreach(archive ${LIB_ARCHIVES})
         if(NOT BUILD_NATIVE)
-            target_compile_options(${LIB_NAME}-${lib_arch} PUBLIC ${LIB_COMPILER_FLAGS} "-march=${lib_arch}")
+            if(LIB_ARCHIVE_ARCHS_${archive})
+                set(arch_list ${LIB_ARCHIVE_ARCHS_${archive}})
+            elseif(LIB_ARCHIVE_ARCHS)
+                set(arch_list ${LIB_ARCHIVE_ARCHS})
+            else()
+                set(arch_list xs3a)
```

Comment:
Id be temped to have a DEFAULT_ARCH variable at the top somewhere (xs3a hardcoded a couple of times)

Reply:
Done - this makes it much clearer why it falls back through the if-else block.